nemo/collections/nlp/models/dialogue/dialogue_zero_shot_slot_filling_model.py

```diff
@@ -0,0 +1,783 @@
+            predicted_slot_class_batch[subtokens_mask], slot_labels[subtokens_mask]
+        )
+
+        return {
+            'val_loss': val_loss,
+            'bio_slot_tp': self.bio_slot_classification_report.tp,
+            'bio_slot_fn': self.bio_slot_classification_report.fn,
+            'bio_slot_fp': self.bio_slot_classification_report.fp,
+            'slot_similarity_tp': self.slot_similarity_classification_report.tp,
+            'slot_similarity_fn': self.slot_similarity_classification_report.fn,
+            'slot_similarity_fp': self.slot_similarity_classification_report.fp,
+            'overall_slot_tp': self.overall_slot_classification_report.tp,
+            'overall_slot_fn': self.overall_slot_classification_report.fn,
+            'overall_slot_fp': self.overall_slot_classification_report.fp,
+            'bio_slot_preds': bio_slot_preds,
+            'bio_slot_labels': bio_slot_labels,
+            'slot_similarity_preds': slot_similarity_preds,
+            'overall_slot_preds': predicted_slot_class_batch,
+            'slot_labels': slot_labels,
+            'input': input_ids,
+            'subtokens_mask': subtokens_mask,
+        }
+
+    @staticmethod
+    def align_mention_to_tokens(bio_labels, mention_labels):
+        """
+        Align the mentions label to token level based on bio labels.
+        
+        Args:
+            bio_labels: (batch_size*max_token_length) [[1, 0, 0, 1, 2], [1, 2, 2, 0, 0]]
+            mention_labels: [[2, 4, 0, 0, 0], [52, 0, 0, 0, 0]]
+        
+        Returns:
+            predicted_slot_class: [[2, 0, 0, 4, 4], [52, 52, 52, 0, 0]]
+        """
+        predicted_slot_class = []
+        for one_bio_labels, one_mention_labels in zip(bio_labels, mention_labels):
+            start_and_end = DialogueZeroShotSlotFillingModel.get_start_and_end_for_bio(one_bio_labels)
+            one_predicted_slot_class = torch.zeros(len(one_mention_labels))
+
+            for idx, one_start_and_end in enumerate(start_and_end):
+                start, inclusive_end = one_start_and_end
+                for idy in range(start, inclusive_end + 1):
+                    one_predicted_slot_class[idy] = one_mention_labels[idx]
+            predicted_slot_class.append(one_predicted_slot_class)
+
+        predicted_slot_class = torch.stack(predicted_slot_class).to(bio_labels.device)
+        return predicted_slot_class
+
+    @staticmethod
+    def get_start_and_end_for_bio(one_bio_labels):
+        """
+        Getting the start and end of the mention from BIO(0, 1, 2) sequence.
+        Mention is start with B(with label 1) and follow by I(with label 2).
+        Args:
+            one_bio_labels: tensor with size (max_token_length, )
+            eg. [1, 0, 0, 1, 2]
+        
+        Returns:
+            start_and_end: list of mention's start and end. 
+            List length is the number of mentions in one_bio_labels.
+            eg. [[0, 0], [3, 4]]
+        """
+        start_and_end = []
```

Comment:
Can you try to read, understand and test this implementation?

`
def get_start_and_end_for_bio(one_bio_labels):
    start_and_end = []
    i = 0
    while i < len(one_bio_labels):
        if one_bio_labels[i] == 1:
            start = i
            i += 1
            while i < len(one_bio_labels) and one_bio_labels[i] == 2:
                i += 1
            exclusive_end = i
            start_and_end.append([start, exclusive_end])
    return start_and_end
`

Reply:
You would need to update the test cases as well as `end` is now `exclusive end `instead of `inclusive end`

Reply:
`
def get_start_and_end_for_bio(one_bio_labels):
    start_and_end = []
    i = 0
    while i < len(one_bio_labels):
        # if encounter the first B (with label 1), 
        # increment counter while encountering I (with label 2)
        if one_bio_labels[i] == 1:
            start = i
            i += 1
            while i < len(one_bio_labels) and one_bio_labels[i] == 2:
                i += 1
            exclusive_end = i
            start_and_end.append([start, exclusive_end])
        # if encounter O (with label 0) or I (with label 2) without a preceding B (with label 1)
        # increment counter
        elif one_bio_labels[i] in [0, 2]:
            i += 1
    return start_and_end
`